django/db/models/sql/compiler.py

```diff
@@ -439,6 +443,15 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
                     raise NotSupportedError('{} is not supported on this database backend.'.format(combinator))
                 result, params = self.get_combinator_sql(combinator, self.query.combinator_all)
             else:
+                distinct_fields = self.get_distinct()
+
+                # This must come after 'select', 'ordering', and 'distinct'
+                # (see docstring of get_from_clause() for details).
+                from_, f_params = self.get_from_clause()
+
+                where, w_params = self.compile(self.where) if self.where is not None else ("", [])
+                having, h_params = self.compile(self.having) if self.having is not None else ("", [])
+
```

Comment:
Yes. `EmptyResultSet` exception is raised here, therefore I decided to move these variables inside the `if` clause where they are used.